### DISCUSSION TITLE
Detect socket remote shutdown

### DISCUSSION THREAD

**Discussion by 13hakta** [resetius/coroio#40] [General]
Hello.
I would like to use your library for my software. I need to use epoll as a socket event reader, so it will be managed with TEPoll.

Can you please explain:
1) I want to start reading/writing socket interaction after some app event, till then socket is just waiting being opened - and here goes next question.
2) how to detect if a socket was closed by remote side without attempts to read or write? With epoll it can be achieved by hooking EPOLLRDHUP event.

**Comment by resetius** [comment]
Hello,
Thank you for your interest in the library! I’m currently quite busy with other work, but I will get back to you with a detailed answer in the next few days.
Thanks for your patience!

**Comment by 13hakta** [comment]
Thanks. I need to implement reverse-reverse-proxy for tcp services (or remoting unix-sockets) and found your implementation as more suitable. I read your articles about this library at LOR/habr and it helped alot to understand more, yet can't understand how to catch events. Спасибо.

**Comment by resetius** [comment]
As I understand it, you'd like to monitor the state of the socket when you're not actively reading from or writing to it (using the ReadSome/WriteSome methods on the Socket). I would suggest adding a new method, perhaps called Monitor, which would trigger when the socket is being closed. What do you think of this idea?

**Reply by 13hakta** [reply]
I think the same way and added through code some additions and added method to Socket:

```
   auto Monitor() {
        struct TAwaitableClose: public TAwaitable<TAwaitableClose> {
            void run() {
                this->ret = true;
            }

            void await_suspend(std::coroutine_handle<> h) {
                this->poller->AddHup(this->fd, h);
            }
        };
        return TAwaitableClose{Poller_,Fd_};
    }

```

to TPollerBase and so on to make it work..
```
    void AddHup(int fd, THandle h) {
        MaxFd_ = std::max(MaxFd_, fd);
        Changes_.emplace_back(TEvent{fd, TEvent::HUP, h});
    }
```
If this is what you propose, then when I finish with my code, I'll send PR to this repo.

I'm still learning coro style and need experienced view.
From point of view of my program flow:
When new connection is accepted, then start handler, but start reading from socket only after trigger event (connection is established on opposite side). I guess that I need to wrap 2 TFuture's in 3rd, wait until it finishes and compare result returned from one of that futures (seen that approach in test).

**Reply by resetius** [reply]
Your implementation of the Monitor method looks solid, and it's in line with what I had in mind.

**Reply by 13hakta** [reply]
> Your implementation of the Monitor method looks solid, and it's in line with what I had in mind.
Can you please review PR?

**Comment by resetius** [comment]
https://github.com/resetius/coroio/pull/41
